### PR TITLE
refactor: opaque gamestate

### PR DESCRIPTION
**Pull request by JonathanStarup**
Relates to #17.

There's a runtime error with records ...

### CODE REVIEW

**Comment by JonathanStarup** [comment]
I'm waiting for debug of enums to find the bug

**Comment by JonathanStarup** [comment]
Can be continued after https://github.com/flix/flix/pull/4705

**Comment by JonathanStarup** [comment]
@mlutze you can look it over now, I've added a hall of bugs to innovations which contains this bug

File: src/Core/GameState.flix [hunk 1]
```diff
@@ -9,86 +10,124 @@ namespace Flixball/Core/GameState {
...
     pub def mkGameState(seed: Int64, strategies: Map[PlayerId, Strategy], b: Board): GameState = {
-        let playerMap = Flixball/Core/Board.players(b) |> List.toMap;
-        let (states, moveLogics) = strategies |> Map.mapWithKey((id, strat) -> {
+        let playerMap = Flixball/Core/Board.players(b);
+        let ais = strategies |> Map.mapWithKey((id, strat) -> {
             let (stateGen, ml) = strat(id);
-            (stateGen(seed), ml)
-        }) |> Utils/Map.unzip;
-        GameState(states, moveLogics, playerMap, b)
+            let pos = Map.get(id, playerMap) |> Utils/Option.unsafeGet;
```
Comment (mlutze) [inline src/Core/GameState.flix:128]: I think every use of an `@Unsafe` function should have a comment explaining what invariant makes it safe.

Also consider: https://github.com/flix/flix/issues/4723

File: src/Core/Board.flix [hunk 2]
```diff
@@ -46,17 +46,19 @@ namespace Flixball/Core/Board {
...
             case _ => None
-        })
+        }) |> List.toMap
     }
 
     pub def player(b: Board, id: PlayerId): Option[Position] =
-        players(b) |>
-            List.find(match (foundId, _) -> foundId == id) |>
-            Option.map(match (_, pos) -> pos)
+        let Board(br) = b;
```
Comment (mlutze) [inline src/Core/Board.flix:58]: Why don't we use `players` anymore?

**Review by mlutze — COMMENTED** [review]
Looks nice, I think this should help keep everything nice and tidy.

File: src/Core/Board.flix (hunk 2 above)
Comment (JonathanStarup) [reply src/Core/Board.flix:58]: For efficiency, I generally think functions with internal access should use internal data. But it probably doesn't matter too much here

**Pull request merged**